grovedb/src/operations/delete.rs

```diff
@@ -22,39 +22,43 @@ impl GroveDb {
         } else {
             let element = self.get_raw(path, &key, transaction)?;
             {
-                let subtrees = match transaction {
-                    None => &mut self.subtrees,
-                    Some(_) => &mut self.temp_subtrees,
-                };
+                let (mut merk, prefix) = self.get_subtrees().get(path, transaction)?;
+                Element::delete(&mut merk, key.clone(), transaction)?;
 
-                let mut merk = subtrees
-                    .get_mut(&Self::compress_subtree_key(path, None))
-                    .ok_or(Error::InvalidPath("no subtree found under that path"))?;
-                Element::delete(merk, key.clone(), transaction)?;
+                // after deletion, if there is a transaction, add the merk back into the hashmap
+                if prefix.is_some(){
+                    self.get_subtrees()
+                        .insert_temp_tree_with_prefix(prefix.expect("confirmed it's some"), merk, transaction);
+                } else {
+                    self.get_subtrees()
+                        .insert_temp_tree(path, merk, transaction);
+                }
             }
 
             if let Element::Tree(_) = element {
                 // TODO: dumb traversal should not be tolerated
                 let mut concat_path: Vec<Vec<u8>> = path.iter().map(|x| x.to_vec()).collect();
                 concat_path.push(key);
                 let subtrees_paths = self.find_subtrees(concat_path, transaction)?;
-                let subtrees = match transaction {
-                    None => &mut self.subtrees,
-                    Some(_) => &mut self.temp_subtrees,
-                };
 
                 for subtree_path in subtrees_paths {
                     // TODO: eventually we need to do something about this nested slices
                     let subtree_path_ref: Vec<&[u8]> =
                         subtree_path.iter().map(|x| x.as_slice()).collect();
                     let prefix = Self::compress_subtree_key(&subtree_path_ref, None);
-                    if let Some(mut subtree) = subtrees.remove(&prefix) {
-                        subtree.clear(transaction).map_err(|e| {
-                            Error::CorruptedData(format!(
-                                "unable to cleanup tree from storage: {}",
-                                e
-                            ))
-                        })?;
+                    let subtree = self
+                        .get_subtrees()
+                        .get_subtree_without_transaction(subtree_path_ref.as_slice());
```

Comment:
perhaps it would be better to add `?` here and do not check `is_ok` with `expect` later unless we want to ignore error for some reason